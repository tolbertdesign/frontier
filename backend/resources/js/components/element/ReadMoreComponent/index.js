import ReadMoreComponent from './ReadMoreComponent.vue'

export default ReadMoreComponent
