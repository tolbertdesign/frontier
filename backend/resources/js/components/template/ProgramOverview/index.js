import ProgramOverview from './ProgramOverview.vue'

export default ProgramOverview
