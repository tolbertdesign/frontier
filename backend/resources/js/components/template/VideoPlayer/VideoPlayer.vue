<template>
  <div>
    <figure class="image is-16by9 mb-4">
      <VideoIframe
        :video-id="activeVideo.external_video_id"
        :source="activeVideo.source"
      />
    </figure>
    <span
      v-if="$store.state.User.is_teacher_user && displayContentLibraryLink"
      class="hidden md:inline absolute content-library-button"
    >
      <ContentLibraryButton />
    </span>

    <BTabs
      v-model="activeTab"
    >
      <BTabItem
        v-for="(videoCollection, index) in videoCollections"
        :key="index"
      >
        <template slot="header">
          <span class="text-xl font-medium inline-block -ml-4">{{ videoCollection.title }}</span>
        </template>
        <VideoGallery
          :videos="videoCollection.videos"
          :selected-video-id="activeVideo.external_video_id"
          @videoSelected="updateActiveVideo"
        />
      </BTabItem>
    </BTabs>

    <div
      v-if="$store.state.User.is_teacher_user"
      class="md:hidden text-center"
    >
      <ContentLibraryButton />
    </div>
  </div>
</template>

<script>
import VideoGallery from '@/components/template/VideoGallery'
import ContentLibraryButton from '@/components/element/ContentLibraryButton'
import VideoIframe from '@/components/element/VideoIframe'

export default {
  name: 'VideoPlayer',
  status: 'prototype',
  release: '1.0',
  components: {
    VideoGallery,
    ContentLibraryButton,
    VideoIframe,
  },
  props: {
    videoCollections: {
      type: Array,
      default: () => [],
    },
    presetVideo: {
      type: Object,
      default: null,
    },
    displayContentLibraryLink: {
      type: Boolean,
      default: true,
    },
  },
  data () {
    return {
      activeTab: 0,
      activeVideoIndex: undefined,
      activeTabKey: undefined,
      selectedVideo: null,
    }
  },
  computed: {
    activeVideo () {
      return this.selectedVideo ? this.selectedVideo : this.presetVideo
    },
    lang () {
      return this.$store.state.lang
    },
  },
  methods: {
    updateActiveVideo (video) {
      this.selectedVideo = video
    },
  },
}
</script>

<style lang="scss">
.b-tabs {
    .tab-content {
        padding: 0;
    }
}
.content-library-button {
    right: 1.5rem;
    margin-top: 0.25rem;
}
</style>
