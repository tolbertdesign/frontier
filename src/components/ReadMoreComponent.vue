<template>
  <div>
    <div
      :class="{ toggled: toggled }"
      class="readmore"
    >
      <span
        v-if="!limitHeight"
        v-html="!toggled && canTruncateBasedOnLimit() ? truncatedText : text"
      />
      <span
        v-if="limitHeight"
        :style="{ height: limit + 'em' }"
        v-html="text"
      />
      <a
        v-if="canTruncateBasedOnLimit()"
        href="javascript:void(0)"
        @click="toggle"
      >
        {{ toggled ? lessText : moreText }}
      </a>
    </div>
  </div>
</template>

<script>
/**
     * example usage
     * <readmore-component text="{{ $testvar }}" more-text="Read More" less-text="Read Less"></readmore-component>
     */
export default {
  name: 'Readmore',
  props: {
    text: { type: String, required: true },
    limit: { type: Number, default: 255 },
    limitHeight: { type: Boolean, default: false },
    moreText: { type: String, default: 'Show More' },
    lessText: { type: String, default: 'Show Less' },
  },
  data () {
    return {
      toggled: false,
    }
  },
  computed: {
    truncatedText () {
      return this.text.substring(0, this.limit) + '...'
    },
  },
  methods: {
    toggle () { this.toggled = !this.toggled },
    canTruncateBasedOnLimit () {
      return this.text.length > this.limit
    },
  },
}
</script>

<style>
.readmore span {
    display: block;
    overflow: hidden;
}

.readmore.toggled span {
    height: auto !important;
}
</style>
