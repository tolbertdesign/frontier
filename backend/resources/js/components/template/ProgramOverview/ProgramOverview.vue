<template>
  <div class="program-overview card-content">
    <VideoPlayer
      v-if="displayVideoPlayer"
      :video-collections="videoCollections"
      :preset-video="presetVideo"
      :display-content-library-link="!program.microsite.hide_character_videos"
      class="pb-8"
    />
    <div class="pb-2">
      <h2 class="text-lg sm:text-2xl font-semibold">
        {{ lang.header_overview_text_override }}
      </h2>

      <div
        class="content"
        v-html="programDescription"
      />
    </div>
  </div>
</template>

<script>
import VideoPlayer from '@/components/template/VideoPlayer'
import axios from 'axios'

export default {
  name: 'ProgramOverview',
  status: 'prototype',
  release: '1.0',
  components: { VideoPlayer },
  props: {
    program: {
      type: Object,
      default: null,
    },
  },
  data () {
    return {
      displayVideoPlayer: false,
      videosRetrieved: false,
      videoCollections: [],
      presetVideo: {},
    }
  },
  computed: {
    lang () {
      return this.$store.state.lang
    },
    programVideos () {
      return {
        name: 'program',
        title: this.lang.program_videos_title,
        url: `/v3/api/videos/program/${this.program.participants[0].id}`,
      }
    },
    characterVideos () {
      return {
        name: 'character',
        title: this.characterVideosTitle,
        url: `/v3/api/videos/character/${this.program.id}`,
      }
    },
    characterVideosTitle () {
      return this.$store.state.User.is_teacher_user ? this.lang.teacher_character_videos_title : this.lang.parent_character_videos_title
    },
    videoCategoriesArray () {
      const videoArray = [this.programVideos]

      if (!this.program.microsite.hide_character_videos) {
        this.addCharacterVideos(videoArray)
      }

      return videoArray
    },
    programOverviewText () {
      return this.program.microsite.overview_text_override
    },
    hasProgramOverviewText () {
      return typeof this.programOverviewText === 'string' && this.programOverviewText.trim() !== ''
    },
    defaultTextOverride () {
      return this.parseLanguage(
        this.$store.state.lang.overview_text_override, {
          event_name: this.program.event_name,
          unit_plural: this.program.unit.name_plural,
        },
      )
    },
    programDescription () {
      if (this.hasProgramOverviewText) {
        return this.programOverviewText
      }

      return this.defaultTextOverride
    },
  },
  created () {
    this.fetchVideos()
  },
  methods: {
    addCharacterVideos (videos) {
      this.$store.state.User.is_teacher_user ? videos.unshift(this.characterVideos) : videos.push(this.characterVideos)
    },
    fetchVideos () {
      axios.all(this.videoCategoriesArray.map(videoCategory => {
        return axios.get(videoCategory.url).then(response => {
          if (response.data && response.data.length > 0) {
            videoCategory.videos = response.data
            this.videosRetrieved = true
          }
        })
      })).then(response => {
        if (this.videosRetrieved) {
          this.setVideos()
        }
      }).catch(error => {
        console.error(error)
      })
    },
    setVideos () {
      this.videoCollections = this.videoCategoriesArray
      this.presetVideo = this.videoCollections[0].videos[0]
      this.displayVideoPlayer = true
    },
  },
}
</script>

<style lang="scss">
    .program-overview .readmore-wrapper {
        .toggled p {
            margin-bottom: 1em;
        }
        .toggled span > :last-child {
            margin-bottom: 0em;
        }
        ol, ul {
            padding-left: 40px;
            margin-top: 1em;
            margin-bottom: 1em;

            ol, ul {
                margin-top: 0em;
                margin-bottom: 0em;
            }
        }
        ul {
            list-style-type: disc;
        }
    }
</style>
