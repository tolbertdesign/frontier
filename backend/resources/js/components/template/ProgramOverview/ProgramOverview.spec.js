import { shallowMount, createLocalVue } from '@vue/test-utils'
import ProgramOverview from './ProgramOverview.vue'
import Buefy from 'buefy'
import VideoPlayer from '@/components/template/VideoPlayer'
import ReadMoreComponent from '@/components/element/ReadMoreComponent'

jest.mock('axios')
const parseLanguage = jest.fn(input => input)

const localVue = createLocalVue()
localVue.use(Buefy, {
  defaultIconPack: 'fas',
})

const defaultIsTeacherUser = true
const defaultHideCharacterVideos = true
const namePluralDefault = 'name plural'
const eventNameDefault = 'event name'
const micrositeOverviewTextOverrideDefault = 'microsite overview text override'
const langOverviewTextOverrideDefault = 'lang overview text override'
const parentCharacterVideosTitle = 'Parent Character Videos'
const teacherCharacterVideosTitle = 'Teacher Character Videos'

const program = {
  id: 1,
  event_name: eventNameDefault,
  unit: {
    name_plural: namePluralDefault,
  },
  microsite: {
    overview_text_override: micrositeOverviewTextOverrideDefault,
    hide_character_videos: defaultHideCharacterVideos,
  },
  participants: [
    { id: 2 },
  ],
}

const $store = {
  state: {
    User: {
      is_teacher_user: defaultIsTeacherUser,
      programs: [program],
    },
    lang: {
      overview_text_override: langOverviewTextOverrideDefault,
      teacher_character_videos_title: teacherCharacterVideosTitle,
      parent_character_videos_title: parentCharacterVideosTitle,
    },
  },
}

describe('ProgramOverview', () => {
  it('Snapshop test for ProgramOverview', () => {
    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot()
    })
  })
  it('has correct length for video collection when character videos are hidden', () => {
    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.videoCategoriesArray.length).toBe(1)
  })
  it('has character videos title and length for videos array when character videos are not hidden for a teacher', () => {
    program.microsite.hide_character_videos = false

    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.videoCategoriesArray.length).toBe(2)
    expect(wrapper.vm.videoCategoriesArray[0].title).toBe(teacherCharacterVideosTitle)
    program.microsite.hide_character_videos = defaultHideCharacterVideos
  })
  it('has character videos title and length for videos array when character videos are not hidden for a parent', () => {
    program.microsite.hide_character_videos = false
    $store.state.User.is_teacher_user = false

    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.videoCategoriesArray.length).toBe(2)
    expect(wrapper.vm.videoCategoriesArray[1].title).toBe(parentCharacterVideosTitle)
    program.microsite.hide_character_videos = defaultHideCharacterVideos
    $store.state.User.is_teacher_user = true
  })
  it('has correct title for character videos for teacher users and a correct program description', () => {
    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.characterVideosTitle).toBe(teacherCharacterVideosTitle)
    expect(wrapper.vm.programDescription).toBe(micrositeOverviewTextOverrideDefault)
  })
  it('has correct title for character videos for parent users', () => {
    $store.state.User.is_teacher_user = false
    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.characterVideosTitle).toBe(parentCharacterVideosTitle)
    $store.state.User.is_teacher_user = defaultIsTeacherUser
  })
  it('has correct program description when program does not have a description', () => {
    program.microsite.overview_text_override = null
    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.programDescription).toBe(langOverviewTextOverrideDefault)
    program.microsite.overview_text_override = micrositeOverviewTextOverrideDefault
  })
  it('adds character videos to correct location in array for teacher', () => {
    const stubbedValue = 'programVideoStub'
    const videoArray = [
      stubbedValue,
    ]

    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    wrapper.vm.addCharacterVideos(videoArray)
    expect(videoArray.length).toBe(2)
    expect(videoArray[0]).not.toBe(stubbedValue)
    expect(videoArray[1]).toBe(stubbedValue)
  })
  it('adds character videos to correct location in array for parent', () => {
    $store.state.User.is_teacher_user = false
    const stubbedValue = 'programVideoStub'
    const videoArray = [
      stubbedValue,
    ]

    const wrapper = shallowMount(ProgramOverview, {
      localVue,
      components: {
        VideoPlayer,
        ReadMoreComponent,
      },
      mocks: {
        $store: $store,
        parseLanguage: parseLanguage,
      },
      propsData: {
        program: program,
      },
    })
    wrapper.vm.addCharacterVideos(videoArray)
    expect(videoArray.length).toBe(2)
    expect(videoArray[0]).toBe(stubbedValue)
    expect(videoArray[1]).not.toBe(stubbedValue)
    $store.state.User.is_teacher_user = true
  })
})
