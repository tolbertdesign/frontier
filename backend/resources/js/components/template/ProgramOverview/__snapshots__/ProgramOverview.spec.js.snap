// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgramOverview Snapshop test for ProgramOverview 1`] = `
<div class="program-overview card-content">
  <!---->
  <div class="pb-2">
    <h2 class="text-lg sm:text-2xl font-semibold">

    </h2>
    <div class="content">microsite overview text override</div>
  </div>
</div>
`;
